src/parser/token.rs

```diff
@@ -0,0 +1,167 @@
+use std::fmt;
+use std::fmt::Formatter;
+use crate::parser::lexer::{Command, DataType, LexInput, LexResult, Modifier, Operator};
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct TokenWithLocation {
+    token: Token,
+    line: u32,
+    column: usize,
+}
+
+impl PartialEq<Token> for TokenWithLocation {
+    fn eq(&self, other: &Token) -> bool {
+        &self.token == other
+    }
+}
+
+impl TokenWithLocation {
+    pub fn as_token(&self) -> &Token {
+        &self.token
+    }
+
+    pub fn into_token(self) -> Token {
+        self.token
+    }
+
+    pub fn line(&self) -> u32 {
+        self.line
+    }
+
+    pub fn column(&self) -> usize {
+        self.column
+    }
+}
+
+impl nom::InputLength for TokenWithLocation {
+    fn input_len(&self) -> usize {
+        // All tokens take up exactly one place in the input token stream
+        self.as_token().input_len()
+    }
+}
+
+pub(crate) fn token_with_location<'i, E, P>(
+    mut parser: P,
+) -> impl FnMut(LexInput<'i>) -> LexResult<'i, TokenWithLocation, E>
+    where
+        P: nom::Parser<LexInput<'i>, Token, E>,
+        E: nom::error::ParseError<LexInput<'i>>,
+{
+    move |input| {
+        let line = input.location_line();
+        // TODO: naive_get_utf8_column might be faster for shorter lines
+        let column = input.get_utf8_column();
+        // Using this syntax because map(parser, || ...)(input) has lifetime issues for parser.
+        parser.parse(input).map(|(leftover, token)| {
+            (
+                leftover,
+                TokenWithLocation {
+                    token,
+                    line,
+                    column,
+                },
+            )
+        })
+    }
+}
+
+#[derive(Clone, PartialEq)]
+pub enum Token {
+    As,
+    Colon,
+    Comma,
+    Command(Command),
+    Comment(String),
+    // Constant(Constant),
```

Comment:
I know this was pre-existing, but since we're reworking this section please clean it out. I can't recall _exactly_ but my guess is that I changed from lexing `i` and `pi` as tokens to just treating them as identifiers and then special-casing them in parsing.